scripts/ontology_to_ontology_mapping_tool/ontology_to_ontology_mapping_tool.py

```diff
@@ -0,0 +1,171 @@
+# Copyright (c) 2019 Memorial Sloan-Kettering Cancer Center.
+#
+# This library is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
+# MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
+# documentation provided hereunder is on an "as is" basis, and
+# Memorial Sloan-Kettering Cancer Center
+# has no obligations to provide maintenance, support,
+# updates, enhancements or modifications.  In no event shall
+# Memorial Sloan-Kettering Cancer Center
+# be liable to any party for direct, indirect, special,
+# incidental or consequential damages, including lost profits, arising
+# out of the use of this software and its documentation, even if
+# Memorial Sloan-Kettering Cancer Center
+# has been advised of the possibility of such damage.
+
+import sys
+import os
+import io
+import argparse
+import pandas as pd
+import requests
+
+ONTOLOGY_MAPPINGS_FILE_URL = "https://raw.githubusercontent.com/cBioPortal/oncotree/master/scripts/ontology_to_ontology_mapping_tool/ontology_mappings.txt"
+VALID_ONCOTREE_CODES_URL = "https://raw.githubusercontent.com/cBioPortal/oncotree/master/resources/resource_uri_to_oncocode_mapping.txt"
+ACCEPTED_ONTOLOGIES = ['ONCOTREE_CODE','UMLS_CODE','NCIT_CODE','ICDO_CODE','HEMEONC_CODE']
+NULL_VALUES = ['','NA']
+
+def add_comments_column_and_log_data(mapped_data,target_file,source_code,target_code,source_file):
+    comments = []
+    completely_resolved_codes = {}
+    ambiguous_codes = {}
+    unrecognized_codes = []
+    many_to_one_codes = {}
+    mapped_codes_count = 0
+    unmapped_codes_count = 0
+
+    columns_to_groupby = mapped_data.columns[:len(mapped_data.columns)-1]
+    grouped_data = mapped_data.groupby(list(columns_to_groupby),sort=False)[target_code].unique().apply(', '.join).reset_index()
+    oncotree_codes_list = pd.read_csv(io.StringIO(requests.get(VALID_ONCOTREE_CODES_URL).content.decode('utf-8')),sep='\t',header=None,keep_default_na=False)
+    valid_oncotree_codes = oncotree_codes_list[2].str.upper().tolist()
```

Comment:
This is not valid. This file contains "triplets" of the form "subject verb object". This approach takes all objects as valid oncotree codes, but sometimes objects are not .. such as with the triplet
`ONC000688	hasPrecursor	ONC000360`
This says that concept ONC000688 has a precursor concept ONC000360 ... and ONC000360 is not a valid oncotree code. Maybe you could only use triplets of the form (subject hasCode object) and then use the object.
Also .. there are some rescinded oncotree codes, such as RESCINDED_URI_PREVIOUSLY_USED_FOR_PMBL. These should never occur in user files.
Also .. there are some confusing cases in the history, where oncotree codes were changed and/or rescinded. So
ONC000376	hasCode	PTCL
ONC000379	hadCode	PTCLNOS
ONC000379	hasRevocation	ONC000376
so this will take both PTCL and PTCLNOS as valid oncotree codes ... but above you give a mapping for PTCL (which is no longer a current code) and do not give a mapping for PTCLNOS (which replaced PTCL) ... there are other cases where codes were rescinded like this. I'm not sure what scenarios we want to support, but it can get complicated.

Reply:
Thanks for pointing it out @sheridancbio. Updated the script to use only the records with `hasCode`.

Reply:
ok, looks good. There still may be some confusion and lack of support if they have not updated to the latest oncotree version .. but that is reasonable